File: src/main/java/com/mrkirby153/snowsgivingbot/entity/GiveawayEntrantEntity.java
package com.mrkirby153.snowsgivingbot.entity;

import lombok.*;

import javax.persistence.*;

@Table(name = "entrants")
@Entity
@AllArgsConstructor
@NoArgsConstructor
@RequiredArgsConstructor
@Data
public class GiveawayEntrantEntity {

    @Id
    @GeneratedValue
    private long id;

    @ManyToOne
    @JoinColumn(name = "giveaway_id")
    @NonNull
    private GiveawayEntity giveaway;

    @Column(name = "user_id")
    private String userId;
}

File: src/main/kotlin/com/mrkirby153/giveaways/jpa/GiveawayEntrantEntity.kt
package com.mrkirby153.giveaways.jpa

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import org.springframework.data.jpa.repository.JpaRepository

/**
 * Entity representing an entrant into a giveaway
 */
@Table(name = "entrants")
@Entity
@LazyEntity
class GiveawayEntrantEntity(
    @Column(name = "user_id")
    val userId: String
) : AutoIncrementingJpaEntity<Long>() {

    @ManyToOne
    @JoinColumn(name = "giveaway_id")
    lateinit var giveaway: GiveawayEntity

    constructor(giveaway: GiveawayEntity, userId: String) : this(userId) {
        this.giveaway = giveaway
    }

}


interface EntrantRepository : JpaRepository<GiveawayEntrantEntity, Long> {

    fun existsByGiveawayAndUserId(giveaway: GiveawayEntity, userId: String): Boolean

    fun getAllByGiveaway(giveaway: GiveawayEntity): List<GiveawayEntrantEntity>
}